float32 velocity
float32 distance
float32 angle